[Tan Sze Ying] iP

tsy24 opened the pull request:
# DukeBot #
> "Hello I'm Duke!
> What can I do for you today?"

DukeBot helps you to keep track of the tasks you need to do. It's
* Text-based
* Easy to use
* ~~FAST~~ SUPER FAST  

All you need to do is
1. Download it [here](https://github.com/tsy24/ip/releases/download/A-jar/ip.jar)
2. Run the .jar file
3. Start adding your tasks!

And it is **FREE**🥳

## Features Available ##
- [x] Add tasks: todo, deadline, event
- [x] List out tasks
- [x] Marking tasks as complete
- [x] Filtering tasks by date or keywords
- [x] Save list of tasks in local file
- [ ] Reminders (coming soon)


If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```javascript    
public static void main(String[] args) {
        Duke duke = new Duke("data/duke.txt");
        duke.run();
}
```

Commit:
Add Gradle support

tsy24 committed:
Implement skeletal version of Duke

tsy24 committed:
Add list ability

tsy24 committed:
Add ability to mark tasks as done

tsy24 committed:
Add support for tracking three types of tasks

tsy24 committed:
Write testcases

tsy24 committed:
Add error handling for wrong inputs

tsy24 committed:
Add delete task ability

tsy24 committed:
Add tasktype enum

tsy24 committed:
Add ability to save tasks to hard disk

tsy24 committed:
Add ability to understand dates

tsy24 committed:
Merge branch 'branch-Level-7'

tsy24 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

tsy24 committed:
Refactor code to extract out closely related code as classes

tsy24 committed:
Organize classes into packages

tsy24 committed:
Add JUnit tests

tsy24 committed:
Fix directory creation for file path

tsy24 committed:
Add JavaDoc

tsy24 committed:
Tweak code to comply with coding standard

tsy24 committed:
Add ability to find a task by searching for a keyword

tsy24 committed:
Merge branch 'branch-A-JavaDoc'

tsy24 committed:
Merge branch 'branch-A-CodingStandard'

tsy24 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java

Superbestron commented on src/main/java/DeadlineTest.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    @Test
+    public void stringConversion_doneTask_indicated() {
```
Following the three part format "featureUnderTest_testScenario_expectedBehavior()" for writing names for test functions, you might want to rename `stringConversion_doneTask_indicated()`to make it clearer

Superbestron commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * DukeException class handles exceptions for Duke.
+ */
+public class DukeException extends Exception{
```
Minor error here: You forgot the space between `Exception` and `{`.

Superbestron commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,125 @@
+     * @return Deadline description string.
+     * @throws DukeException If description is empty or does not contain a date.
+     */
+    public static String checkDeadlineDescription(String des) throws DukeException {
+        String description = des.trim();
+        if (description.isEmpty()) {
+            throw new DukeException("The description of a deadline cannot be empty!\n");
+        } else if (!description.contains(" /by ")) {
+            throw new DukeException("The new deadline is missing a date!\n");
+        } else {
+            return description;
+        }
+    }
```
I think it's great that you wrote exceptions for a number of invalid inputs.
However, the logic is the same for `Deadline`, `Event` and `Todo`. Perhaps you might want to consider using inheritance/polymorphism instead via restructuring the switch case above to achieve the same purpose?

Superbestron commented on src/main/java/duke/command/EditCommand.java:
```diff
@@ -0,0 +1,54 @@
+/**
+ * EditCommand class handles the commands that edit a particular task in the tasklist.
+ */
+public class EditCommand extends Command {
```
Perhaps it would be better if you split this class into DoneCommand and DeleteCommand because they themselves achieve slightly different functionalities? 👍

chrus-chong commented on src/main/java/DeadlineTest.java:
```diff
@@ -0,0 +1,36 @@
+    public void testFileFormatConversion() {
+        Deadline deadline = new Deadline("project", LocalDate.of(2021, 9, 1));
+        assertEquals("D / 0 / project / 2021-09-01", deadline.toFileFormat());
+    }
```
In line with the 3 part format under the testing segment of the [Java Coding Standard](https://nus-cs2103-ay1920s2.github.io/website/coding-standards/java/intermediate.html), you can consider renaming your method from `testFileFormatConversion( )` to `fileFormatConversion_deadlineTask_stringReturned( )` to enhance readability.

chrus-chong commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,55 @@
+     */
+    public void run() {
+        ui.showWelcome();
+        boolean isExit = false;
```
You may consider naming your boolean variables to sound more intuitive. Perhaps I can recommend using `hasExited `instead?

chrus-chong commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,95 @@
+                        Event event = new Event(parameters[2], LocalDate.parse(parameters[3]));
+                        event.setDone(Integer.parseInt(parameters[1]));
+                        taskList.add(event);
+                        break;
```
Should a default statement be added here to deal with the off-chance of parameters[0] corresponding to none of the above cases?

chrus-chong commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,43 @@
+     * Prints the welcome message.
+     */
+    public void showWelcome() {
+        String startText = "Hello! I'm duke.Duke\n" + "What can I do for you?";
```
Could this be a typo? Did you mean to say `Hello! I'm Duke\n` instead?

chrus-chong commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,59 @@
+    @Override
+    public void execute(TaskList taskList, Ui ui, Storage storage) throws DukeException {
+        String messageHeader = "Alright! New task added:\n";
+        switch (type) {
+        case DEADLINE:
+            Deadline deadline = new Deadline(parameters[0].trim(), LocalDate.parse(parameters[1]));
+            taskList.addTask(deadline);
+            String deadlineMessage = messageHeader + deadline + taskList.getListStatus();
+            ui.printMessage(deadlineMessage);
+            break;
+        case EVENT:
+            Event event = new Event(parameters[0], LocalDate.parse(parameters[1]));
+            taskList.addTask(event);
+            String eventMessage = messageHeader + event + taskList.getListStatus();
+            ui.printMessage(eventMessage);
+            break;
+        case TODO:
+            ToDo toDo = new ToDo(parameters[0]);
+            taskList.addTask(toDo);
+            String toDoMessage = messageHeader + toDo + taskList.getListStatus();
+            ui.printMessage(toDoMessage);
+            break;
```
Should a default statement be added here in case of an unexpected error and type does not match the above cases?

chrus-chong reviewed (commented):
Nicely done! I can't seem to find any recurring major coding standard violations. Keep up the good work!

chrus-chong commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,64 @@
+    public String toFileFormat() {
+        return String.format("%d / %s", isDone ? 1 : 0, description);
+    }
+}
```
Should this method be renamed to `convertToFileFormat` in line with the Java Coding Standard that all names representing methods should be verbs?

chrus-chong commented on src/main/java/duke/command/EditCommand.java:
```diff
@@ -0,0 +1,54 @@
+                break;
+            }
+        }
+    }
```
In line with the Java Coding Standard, descriptive comments should be made for all public methods. Perhaps stating that the execute method in the parent class "Executes itself" may not be sufficient to explain the unique behaviour of `execute` in each type of Command?

chrus-chong commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,97 @@
+    public String getTaskList() {
+        String list = "These are the tasks in your list:\n";
+        for (int i = 0; i < tasks.size(); i++) {
+            list += (i + 1) + ". " + tasks.get(i).toString() + "\n";
```
Should line breaks be used here to facilitate ease of reading?
Perhaps you can consider adding a break before the `+` operator in line with the [Java Coding Standard](https://nus-cs2103-ay1920s2.github.io/website/coding-standards/java/intermediate.html#variables).

chrus-chong reviewed (commented):
Well done on the project. LGTM, almost. Consider adding Javadocs for all the unique `execute( ) `methods.

tsy24 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

tsy24 committed:
Set up gradle to run unit tests

tsy24 committed:
Add GUI to Duke

tsy24 committed:
Merge branch 'branch-Level-10'

tsy24 committed:
Use checkStyle to detect coding style violations

tsy24 committed:
Merge branch 'branch-A-Checkstyle'

tsy24 committed:
Edit code based on PR review

tsy24 committed:
Add assertions

Add assertions to document important assumptions that should hold
at various points in the code.

tsy24 committed:
Delete Ui class

GUI added handles the UI, making the Ui class redundant.

Let's delete the Ui class and remove the Ui object from Duke class.

Deleting the Ui class will result in better code quality as we
remove dead code.

tsy24 committed:
Remove magic strings

Using magic strings in the code makes it hard for others to
understand the code.

Let's declare the strings as static constants, with names that
describe what the string is for.

Using a named constant makes the code easier to understand because
the name tells us more about the meaning of the string.

tsy24 committed:
Merge pull request #1 from tsy24/branch-A-Assertions

Add assertions

tsy24 committed:
Merge branch 'master' into branch-A-CodeQuality

tsy24 committed:
Merge pull request #2 from tsy24/branch-A-CodeQuality

Improve code quality

tsy24 committed:
Add .yml file for Continuous Integration

tsy24 committed:
Add tagging feature

No means to tag other information to task

Tags can be added with the 'tag' command. Tags are printed with
the task and saved in the local file.

A tagging feature allows users to add additional information
to their task.

tsy24 committed:
Merge branch 'branch-C-Tagging'

tsy24 committed:
Refactor gradle.yml

tsy24 committed:
Update test files and formatting

tsy24 committed:
Edits for checkstyle

tsy24 committed:
Delete I/O redirection style automated testing

tsy24 committed:
Revert "Delete I/O redirection style automated testing"

This reverts commit a852d6aa4058d5500124d657d840b7d3c13a1b68.

tsy24 committed:
Remove I/O redirection style automated testing

tsy24 committed:
Remove I/O redirection style test check in CI

tsy24 committed:
Improve the GUI

Error messages will be displayed in red. Pictures
are displayed in a circle. Change background color to purple.

tsy24 committed:
Finalize features

Change name of bot to Daisy. Give credit to code referenced.

tsy24 committed:
Fix format of response for task list status

tsy24 committed:
Add Ui screenshot and user guide

tsy24 committed:
Add code to close GUI after bye command

tsy24 committed:
Change platform exit to system exit

tsy24 committed:
Remove unused import

tsy24 committed:
Update gradle

Update javaFx version, application and jar name

tsy24 committed:
Set resizable to false
